### DISCUSSION TITLE
Switching from C++-17 to C++-20?

### DISCUSSION THREAD

**Discussion by achimste** [official-stockfish/Stockfish#4869] [Ideas]
Hi chess enthusiasts!

I am playing around with the SF engine for quite some time now and I already switched to C++-20. I see some minor benefits doing so
especially with the introduction of std::popcount. This unifies the code in such a way that we do not need to write different code for different architectures. This is handled by the compiler.

Usage of std::has_single_bit in the update_blocker function. I know it is just one line of code but it improves readability instead of wordy
"b && !more_than_one(b)"

What do you think?

**Comment by Disservin** [comment]
sure... would just require everyone on fishtest to update their compilers which is a bit difficult to enforce.

**Reply by MrBrain295** [reply]
Don't most compilers support both?

**Reply by Disservin** [reply]
well gcc and clang dont support c++20 fully, and earlier compiler versions support even less.. so it could be that we end up writing c++20 code which is supported by the latest compiler versions but fishtest workers are not up to date so it wont work

**Reply by Disservin** [reply]
https://clang.llvm.org/cxx_status.html
c++ 17 since clang 5, c++20 only partially supported

gcc
c++ 20 has experimental support for the latest revision of the C++ standard, which was published in 2020.

**Reply by vondele** [reply]
I think the lowest version of gcc that currently still supports SF compilation is 8.x or so. We need to be considerate of older distros before we up requirements significantly. We have still quite some workers at 8.3 right now (checked).

**Reply by Disservin** [reply]
yeah i think we will kinda have to go with the flow of time and follow the upgrade cycle of the distros which bump the default compiler toolchain every so often, though i dont think we have a need for coroutines/modules

**Comment by Sopel97** [comment]
I don't think support for C++20 is good enough to warrant that.

**Reply by wangjingbo123456** [reply]
I think support
